Modify the check_scan_job module in the scheduler.py

xiaoruiDong opened the pull request:
Major changes:
(1) Since the `if job.job_status[1] != 'done'` is checked at the beginning, the `invalidation_reason = crash` will never  triggered. So I depreciate the relative codes
(2) The invalidation of a rotor with a barrier larger than 40 kJ should be triggered after the 'lowest-conformer' check.
(3) The previous code/logic structure is complicated. The workflow is simplified:
```
loop: find corresponding rotor
         if found:
             sequential checks (`break` right after any check is failed)
         else not found:
             raise error
assign invalidation_reason, scan_path if not trsh
plot scan curve if available
```
(4) Scan curve plot also for invalid rotors (distinguished by name)
(5) Scan curve is now dot-line curve instead of solely dots
(6) When the conformer with lower energy is found, directly rotate to the lowest conformer instead of the nearest minima.

-----------------------------------------------------------------------------------
Other thoughts:
(1) Tighten the initial-final difference threshold. As long as, there is no change in other dihedrals. The difference ~ 2 kJ (though not large) is usually caused by a change in dihedral R-O-H between initial and final conformers. 
(2) Tighten the lowest conformer threshold. I find some cases failing to find the lowest, especially when the barrier is not high. The current threshold is to avoid false positives in the check, I guess. However, it is usually not that big (from some of my methyl group's example, ~2% difference). To avoid false positives, it can be checked with symmetries.
(3) I am developing a debugging tool now. The problem is how to implement it. Let me give an example, here are three cases (1) ESS failed because of (bond breaking) (2) ESS converged but has a detectable problem (large jump) (3) ESS converged and the curve looks correct by current checking. But the debug tool finds there is an obvious dihedral change, and it is confirmed by GaussView visualization. The workflow will become less clear then. It will be desirable if it can completely replace the whole check_scan_job module, but then the choice of the outlier threshold should be more tailored. I am implementing 1.5 X IQR as the threshold. Generally looks good given some example cases, but need to test its performance on a larger scale.

alongd commented on arc/plotter.py:
```diff
@@ -178,17 +178,21 @@ def plot_rotor_scan(angle, v_list, path=None, pivots=None, comment=''):
     plt.figure(figsize=(4, 3), dpi=120)
     plt.subplot(1, 1, 1)
-    plt.plot(angle, v_list, 'g.')
+    plt.plot(angle, v_list, 'g.-')
```
Nice. Can you paste in the PR comments a sample image created using this modification?

xiaoruiDong replied:
It looks like this. I personally think that adding lines can help me better observe the curve. It is helpful especially when I am debugging the rotor scan scripts.
![image](https://user-images.githubusercontent.com/35771233/64034706-2ebae080-cb1d-11e9-85ba-8112804c4157.png)

alongd replied:
Great! Perhaps make the dots in a different color

alongd commented on arc/plotter.py:
```diff
@@ -178,17 +178,21 @@ def plot_rotor_scan(angle, v_list, path=None, pivots=None, comment=''):
     plt.ylabel('V (kJ/mol)')
     plt.tight_layout()
     plt.show()
+    if commment:
```
We can remove this block under the `if path is not None and pivots is not None:` since it's only being used if that statement is True

xiaoruiDong replied:
Yes, you are right! I reframed it.

alongd commented on arc/scheduler.py:
```diff
@@ -1746,144 +1745,138 @@ def check_scan_job(self, label, job):
+                    self.species_dict[label].set_dihedral(
+                        pivots=self.species_dict[label].rotors_dict[i]['pivots'],
+                        scan=self.species_dict[label].rotors_dict[i]['scan'],
+                        deg_increment=min_index*job.scan_res)
```
good idea!
(please add a space on both sides of the `*` operator)

xiaoruiDong replied:
Thanks! space added.

alongd commented:
Thanks! @dranasinghe is also working on better rotor troubleshooting methods, perhaps the three of us can meet to decide how best to continue?

xiaoruiDong commented:
Yes, Duminda and I are working on that. Based on the idea from Duminda, I simplified and modified the workflow. I am trying to do a brief data analysis based on the jobs I have and try to diagnosis the behavior of the new debugger. I am free today for discussions

xiaoruiDong commented:
@alongd I update the branch. Once you think it is good, I can squash the fixup commits. Thanks!

alongd commented on arc/scheduler.py:
```diff
@@ -1746,144 +1745,138 @@ def check_scan_job(self, label, job):
+                                                                               pivots=job.pivots)
+                    logger.info(message)
+                    # Find the rotation dihedral in degrees to the closest minimum:
+                    min_index = np.argmin(v_list)
```
👍

alongd commented on arc/scheduler.py:
```diff
@@ -1746,144 +1745,138 @@ def check_scan_job(self, label, job):
+                    break
+
+                # 3. Check the barrier height
+                if (np.max(v_list) - v_list[0]) > maximum_barrier:
```
Let's change ` v_list[0]` to `np.min(v_list)`. I know that by this point we assume there was no lower conformer, but I think it's cleaner and more comprehensive (we could have found a conformer lower by just 1 kJ/mol for which we don't re-run opt)

xiaoruiDong replied:
Thanks! I agree that is a more comprehensive approach. BTW, in the future, what is a desirable behavior of ARC when finding a conformer with dE < 2 kJ/mol? We will switch to that conformer, right? The current status is that ARC does not tell, there is a lower conformer if the difference is within 2 kJ/mol, and I personally rerun the job.

alongd commented on arc/scheduler.py:
```diff
@@ -1746,144 +1745,138 @@ def check_scan_job(self, label, job):
+                        rotor_path=job.local_path_to_output_file, label=label,
+                        pivots=self.species_dict[label].rotors_dict[i]['pivots'])[0]
+                    symmetry = ' has symmetry {0}'.format(self.species_dict[label].rotors_dict[i]['symmetry'])
+                    logger.info('Rotor scan {scan} between pivots {pivots}'
```
let's combine `' has symmetry {0}'` into the `logger.info`

alongd commented on arc/scheduler.py:
```diff
@@ -1746,144 +1745,138 @@ def check_scan_job(self, label, job):
         else:
             raise SchedulerError('Could not match rotor with pivots {0} in species {1}'.format(job.pivots, label))
+
+        if not trsh: # Not append info to troubleshooting scan
```
Please change to `Only continue if not troubleshooting the scan`

alongd replied:
Double-space before inline `#` comment (here and below)

alongd commented on arc/scheduler.py:
```diff
@@ -1746,144 +1745,138 @@ def check_scan_job(self, label, job):
             raise SchedulerError('Could not match rotor with pivots {0} in species {1}'.format(job.pivots, label))
+
+        if not trsh: # Not append info to troubleshooting scan
+            if invalidate: # if success or reset conformer, invalidate = false
```
I didn't understand the comment

xiaoruiDong replied:
That works like a reminder. At this stage, if the rotor looks good, `invalidate = False`; if a lower conformer is found, then `invalidate = False`. Otherwise, `invalidate = True`.   It looks trivial, so I removed it in the new PR.

xiaoruiDong commented:
Thanks @alongd for review the PR! I made changes accordingly. Since most comments are about rephrasing comments, I squashed all the fixup commits together.

alongd commented:
I'm sure you have many exotic rotor scans by now, let's be sure to save theme somewhere to make unit tests out of them when we transition all troubleshooting methods into a separate module.
